### PR TITLE
Bug 1727318 - Dispatcher can't be stopped while it's being shutdown

### PR DESCRIPTION
**Pull request by brizental**
I also found another bug on the rate limit implementaion while working on this bug, which is fixed in 5be95ad.

Detailed descriptions on the commit messages explain the changes made.

### Pull Request checklist ###

- [x] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint` Runs _all_ linters
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] ~**Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one~
   - I think CHANGELOG changes are unnecessary here. This is part of rate limiting implementation, which is already there.
- [ ] ~**Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work~

### CODE REVIEW

File: glean/src/core/dispatcher.ts [hunk 1]
```diff
@@ -320,9 +324,18 @@ class Dispatcher {
...
    * In order to re-start the dispatcher, call the `resume` method.
+   *
+   * # Note
+   *
+   * In case a Shutdown command has been launched before this command,
+   * this command will result in the queue being cleared.
    */
   stop(): void {
-    this.launchInternal({ command: Commands.Stop }, true);
+    if (this.shuttingDown) {
```
Comment (Dexterp37) [inline glean/src/core/dispatcher.ts:339]: Wouldn't this disrupt the sequence of actions when shutting down? Shouldn't we still be executing tasks up until the "shutdown"?

Comment (brizental) [reply glean/src/core/dispatcher.ts:339]: So, the stop command disrupts the sequence of actions, with or without shutdown, by default. It is up to the caller to use `stop(false)` when they don't want stop to be enqueued in front of everything else.

Comment (brizental) [reply glean/src/core/dispatcher.ts:339]: I do agree this could be surprising behaviour, I was torn while implementing.

Comment (Dexterp37) [reply glean/src/core/dispatcher.ts:339]: I see, that's ok, sad but ok :)

File: glean/tests/unit/core/dispatcher.spec.ts [hunk 2]
```diff
@@ -464,6 +464,44 @@ describe("Dispatcher", function() {
...
+      await new Promise<void>(resolve => setTimeout(() => resolve(), 100));
+      dispatcher.stop();
+    });
+
+    let executionCounter2 = 0;
+    const counterTask2 = (): Promise<void> => {
+      executionCounter2++;
+      return Promise.resolve();
+    };
+    // Launch another group of tasks
```
Comment (Dexterp37) [inline glean/tests/unit/core/dispatcher.spec.ts:493]:
```suggestion
    // Launch another group of tasks.
```

File: glean/tests/unit/core/dispatcher.spec.ts [hunk 3]
```diff
@@ -464,6 +464,44 @@ describe("Dispatcher", function() {
...
+    dispatcher = new Dispatcher();
+    dispatcher.flushInit();
+
+    let executionCounter1 = 0;
+    const counterTask1 = (): Promise<void> => {
+      executionCounter1++;
+      return Promise.resolve();
+    };
+
+    // Launch a group of tasks
```
Comment (Dexterp37) [inline glean/tests/unit/core/dispatcher.spec.ts:477]:
```suggestion
    // Launch a group of tasks.
```

**Review by Dexterp37 — APPROVED** [review]
r+ with a changelog entry added.

**Pull request merged**